drgn_tools/kernfs_memcg.py

```diff
@@ -0,0 +1,172 @@
+    """
+    count = 0
+    for kn in for_each_kernfs_node(prog):
+        if kernfs_node_of_memcgroup(kn):
+            count = count + 1
+            path = kernfs_path(kn).decode()
+            print("kernfs_node: ", hex(kn.value_()), "  ", path)
+
+    print("Total number of memcg kernfs_node objects: ", count)
+
+
+def get_num_active_mem_cgroups(prog: Program) -> int:
+    """
+    Get number of active mem cgroups.
+    """
+    mem_cgroup_subsys = prog["cgroup_subsys"][4]
+    # add 1 to number of active memcgroups to account for root memcgroup
+    return mem_cgroup_subsys.root.cgrp.nr_descendants.value_() + 1
+
+
+def get_num_dying_mem_cgroups(prog: Program) -> int:
+    """
+    Get number of inactive or dying mem cgroups.
+    """
+    mem_cgroup_subsys = prog["cgroup_subsys"][4]
+    return mem_cgroup_subsys.root.cgrp.nr_dying_descendants.value_()
+
+
+# By default we scan all pages, that have memcg ref
+# but if max_pages is specified then we bail out
+# after getting those many pages or scanning all
+# pages , whichever happens first
+def dump_page_cache_pages_pinning_cgroups(prog: Program, max_pages: int = 0):
+    """
+    Dump all page-cache pages that have reference to a mem-cgroup.
+    The ouput also contains information such as the cgroup that is pinned, its flags
+    (to indicate current state of cgroup) and file cached by this page.
+
+    :params: max_pages: specify how many pages to find. For default (0) all such pages
+    are listed.
+
+    """
+    PG_slab_mask = 1 << prog.constant("PG_slab")
+    mem_cgroup_root = prog["cgroup_subsys"][4].root
+    total_count = 0
+    found_count = 0
+    for page in for_each_page(prog):
+        total_count = total_count + 1
+        try:
+            # Ignore slab pages
+            if page.flags & PG_slab_mask:
+                continue
+            # Ignore non page-cache pages
+            if not page.mapping:
+                continue
+            try:
+                mem_cgroup = page.mem_cgroup
+            except AttributeError:
+                mem_cgroup = page.memcg_data
+
+            if not mem_cgroup.value_() or mem_cgroup.value_() & 3:
+                continue
+            cgroup_subsys_state = cast(
+                "struct cgroup_subsys_state *", mem_cgroup
+            )
+            if cgroup_subsys_state.cgroup.root == mem_cgroup_root:
+                found_count = found_count + 1
+                cgrp = cgroup_subsys_state.cgroup
+                address_space = page.mapping
+                inode = address_space.host
+                dentry_addr = inode.i_dentry.first.value_() - 0xB0
+                if dentry_addr <= 0:
+                    continue
+                dentry = Object(prog, "struct dentry", address=dentry_addr)
+                print(
+                    f"page: 0x{page.value_():x} cgroup: {cgroup_path(cgrp).decode()} flags: {cgrp.self.flags.value_()} dpath: {dentry_path(dentry.address_of_()).decode()}\n"
```

Comment:
sound like `inode_path` drgn helper can get the path from inode, with that, you can remove the logic to locate dentry.

Again please make that flags human readable

Reply:
Done